plugins/woocommerce/src/Internal/ProductQueryFilters/FilterClausesGenerator.php

```diff
@@ -254,6 +112,96 @@ private function price_filter_clauses( $args, $wp_query ) {
 		return $args;
 	}
 
+	/**
+	 * Add query clauses for attribute filter.
+	 *
+	 * @param array $args              Query args.
+	 * @param array $chosen_attributes {
+	 *     Chosen attributes array
+	 *
+	 *     @type array {$taxonomy: Attribute taxonomy name} {
+	 *         @type string[] $terms      Chosen terms' slug.
+	 *         @type string   $query_type Query type. Accepts 'and' or 'or'.
+	 *     }
+	 * }
+	 * @return array
+	 */
+	public function add_attribute_clauses( $args, $chosen_attributes ) {
+		if ( empty( $chosen_attributes ) ) {
+			return $args;
+		}
+
+		global $wpdb;
+
+		// The extra derived table ("SELECT product_or_parent_id FROM") is needed for performance
+		// (causes the filtering subquery to be executed only once).
+		$clause_root = " {$wpdb->posts}.ID IN ( SELECT product_or_parent_id FROM (";
+		if ( 'yes' === get_option( 'woocommerce_hide_out_of_stock_items' ) ) {
+			$in_stock_clause = ' AND in_stock = 1';
+		} else {
+			$in_stock_clause = '';
+		}
+
+		$attribute_ids_for_and_filtering = array();
+
+		foreach ( $chosen_attributes as $taxonomy => $data ) {
+			$all_terms                  = get_terms( $taxonomy, array( 'hide_empty' => false ) );
+			$term_ids_by_slug           = wp_list_pluck( $all_terms, 'term_id', 'slug' );
+			$term_ids_to_filter_by      = array_values( array_intersect_key( $term_ids_by_slug, array_flip( $data['terms'] ) ) );
+			$term_ids_to_filter_by      = array_map( 'absint', $term_ids_to_filter_by );
+			$term_ids_to_filter_by_list = '(' . join( ',', $term_ids_to_filter_by ) . ')';
+			$is_and_query               = 'and' === strtolower( $data['query_type'] );
+
+			$count = count( $term_ids_to_filter_by );
+
+			if ( 0 !== $count ) {
+				if ( $is_and_query && $count > 1 ) {
+					$attribute_ids_for_and_filtering = array_merge( $attribute_ids_for_and_filtering, $term_ids_to_filter_by );
+				} else {
+					$clauses[] = "
+							{$clause_root}
+							SELECT product_or_parent_id
+							FROM {$this->get_lookup_table_name()} lt
+							WHERE term_id in {$term_ids_to_filter_by_list}
+							{$in_stock_clause}
+						)";
+				}
+			}
+		}
+
+		if ( ! empty( $attribute_ids_for_and_filtering ) ) {
+			$count                      = count( $attribute_ids_for_and_filtering );
+			$term_ids_to_filter_by_list = '(' . join( ',', $attribute_ids_for_and_filtering ) . ')';
+			$clauses[]                  = "
+				{$clause_root}
+				SELECT product_or_parent_id
+				FROM {$this->get_lookup_table_name()} lt
+				WHERE is_variation_attribute=0
+				{$in_stock_clause}
+				AND term_id in {$term_ids_to_filter_by_list}
+				GROUP BY product_id
+				HAVING COUNT(product_id)={$count}
+				UNION
+				SELECT product_or_parent_id
+				FROM {$this->get_lookup_table_name()} lt
+				WHERE is_variation_attribute=1
+				{$in_stock_clause}
+				AND term_id in {$term_ids_to_filter_by_list}
+				GROUP BY product_or_parent_id
+				HAVING COUNT(product_or_parent_id)>={$count}
```

Comment:
@Konamiman I have to add these line to fix the filtering with variable products. Please see https://github.com/woocommerce/woocommerce/pull/43772/commits/06bb80f0b6c9ae52335656bac69cf641001b67fb for more info.

The same happens to main query:
![image](https://github.com/woocommerce/woocommerce/assets/5423135/2c9eeeed-0a95-469e-9998-836e17bb6ab1)

We should also fix it in `Filterer`, but let's do it in a separated PR.

Reply:
The condition here is not correct, I fixed it in https://github.com/woocommerce/woocommerce/pull/43772/commits/c78e84414db423902b79d44e9ad973e036520aae. We should fix the same logic in `Filterer` because the products showing for `query_type_<taxonomy>=and` is incorrect.

Reply:
cc @Konamiman